src/cls/rbd/cls_rbd.cc

```diff
@@ -4428,6 +4514,89 @@ int mirror_instances_remove(cls_method_context_t hctx, bufferlist *in,
 }
 
 /**
+ * Input:
+ * @param start_after: key to start after
+ * @param max_return: max return items
+ *
+ * Output:
+ * @param std::map<std::string, cls::rbd::MirrorImageMap>: image mapping
+ * @returns 0 on success, negative error code on failure
+ */
+int mirror_image_map_list(cls_method_context_t hctx, bufferlist *in,
+                          bufferlist *out) {
+  std::string start_after;
+  uint64_t max_return;
+  try {
+    bufferlist::iterator it = in->begin();
+    ::decode(start_after, it);
+    ::decode(max_return, it);
+  } catch (const buffer::error &err) {
+    return -EINVAL;
+  }
+
+  std::map<std::string, cls::rbd::MirrorImageMap> image_mapping;
+  int r = mirror::mirror_image_map_list(hctx, start_after, max_return, &image_mapping);
+  if (r < 0) {
+    return r;
+  }
+
+  ::encode(image_mapping, *out);
+  return 0;
+}
+
+/**
+ * Input:
+ * @param std::map<std::string, cls::rbd::MirrorImageMap>: image mapping
```

Comment:
I am still thinking that these two methods (set and remove) should just take two parameters (global image id and cls::rbd::MirrorImageMap) instead of a map. The caller of these methods can just execute the method multiple times within the same operation. In terms of OSD complexity, you have a little extra overhead for multiple exec calls in the operation, but I think it will simplify the update state machine and it mimics the behavior of all other RBD cls methods.